Refactor VMManager out of VMInspector.

MenloDorian opened the pull request:
#### ccf0be33dc3b2f7d60c4cde045955933012e8101
<pre>
Refactor VMManager out of VMInspector.
<a href="https://bugs.webkit.org/show_bug.cgi?id=297578">https://bugs.webkit.org/show_bug.cgi?id=297578</a>
<a href="https://rdar.apple.com/158659702">rdar://158659702</a>

Reviewed by Keith Miller and Yijia Huang.

Previously, we only needed to enumerate VMs for debugging purposes.  As a result, we implemented
a container of VMs in VMInspector to support enumeration and inspection.  However, since then,
production code has come to rely on this container of VMs.  Additionally, we will soon need to
add more code in the area of VM management and coordination.  Hence, it makes sense to refactor
this container code out of VMInspector into a VMManager class of its own.

Changes include:

1. Changing VMManager::forEachVM() to take a ScopedLambda instead of a Function.  forEachVM() is
   an iteration mechanism, and the life cycle of the callback does not need to extend beyond the
   iteration operation.  Hence, then context of the iteration function can live on the stack in
   a ScopedLambda instead of needing to be malloc&apos;ed in a Function.

2. Some clients were previously using VMInspector&apos;s VM iteration simply for the purpose of
   finding a VM matching some criteria.  Most of the time, the recent most added or searched
   for VM is the matching one.  Hence, we have a s_recentVM cache for that.  However, the
   iteration operation should still be holding the g_vmListLock to prevent the VM instance from
   being destructed before we call the client&apos;s callback with the VM in s_recentVM.

   Previously, this was done by requiring the client to manually acquire g_vmListLock before
   checking s_recentVM, and subsequently, using VMInspector::iterate() to iterate the VMs.
   This can be a source of bugs if the client does handle the lock correctly.  Additionally,
   it requires a lot more of VM container implementation to be exposed as API.

   In VMManager, we introduce a findVM() operation that will encapsulate this synchronization,
   checking of s_recentVM, and thereafter, iterating the container if needed.  This makes the
   interface more robust and simpler to use correctly.

3. Removed the dependency for VMInspector::isValidExecutableMemory() and
   VMInspector::codeBlockForMachinePC() on the VMInspector&apos;s lock.

   VMInspector::isValidExecutableMemory() never needed the VMInspector lock.  The real lock that
   actually needs to be acquired belongs to the ExecutableAllocator instead.
   VMInspector::isValidExecutableMemory() already acquires that.

   VMInspector::codeBlockForMachinePC() now uses VMManager::forEachVM() which handles its
   locking internally.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/disassembler/ARM64/A64DOpcode.cpp:
(JSC::ARM64Disassembler::A64DOpcodeMoveWide::handlePotentialDataPointer):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
* Source/JavaScriptCore/jsc.cpp:
(startTimeoutTimer):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/VMManager.cpp: Added.
(JSC::WTF_REQUIRES_LOCK):
(JSC::VMManager::add):
(JSC::VMManager::remove):
(JSC::VMManager::isValidVMSlow):
(JSC::VMManager::dumpVMs):
(JSC::VMManager::findVMImpl):
(JSC::VMManager::forEachVMImpl):
(JSC::VMManager::forEachVMWithTimeoutImpl):
* Source/JavaScriptCore/runtime/VMManager.h: Added.
(JSC::VMManager::isValidVM):
(JSC::VMManager::findVM):
(JSC::VMManager::forEachVM):
(JSC::VMManager::forEachVMWithTimeout):
* Source/JavaScriptCore/tools/HeapVerifier.cpp:
(JSC::HeapVerifier::checkIfRecorded):
* Source/JavaScriptCore/tools/Integrity.cpp:
(JSC::Integrity::Analyzer::analyzeVM):
* Source/JavaScriptCore/tools/IntegrityInlines.h:
(JSC::Integrity::doAudit):
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::vmForCallFrame):
(JSC::VMInspector::codeBlockForMachinePC):
(JSC::VMInspector::singleton): Deleted.
(JSC::VMInspector::add): Deleted.
(JSC::VMInspector::remove): Deleted.
(JSC::VMInspector::isValidVMSlow): Deleted.
(JSC::VMInspector::dumpVMs): Deleted.
(JSC::VMInspector::forEachVM): Deleted.
* Source/JavaScriptCore/tools/VMInspector.h:
(JSC::VMInspector::isValidVM): Deleted.
(JSC::VMInspector::WTF_RETURNS_LOCK): Deleted.
(JSC::VMInspector::WTF_REQUIRES_LOCK): Deleted.
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::reportToVMsForDestruction):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:

Canonical link: <a href="https://commits.webkit.org/298937@main">https://commits.webkit.org/298937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a86897e9e5f07b71524ab3d949d6184eb261a3e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117249 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123347 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69234 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cc44fcc5-bc5a-4392-ba34-83bec0cdf99a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45506 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1d9ae8bd-a544-43d4-a3b0-0292254d5c4d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29934 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69478 "Passed tests") | | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/28994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67021 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/109354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/99332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126469 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115756 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44146 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/33163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44502 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101355 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97440 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24810 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42789 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44019 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/49678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144456 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/43475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/46820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45171 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -0,0 +1,124 @@
+}
+
+template<typename Functor>
+static void iterateVMs(const Functor& functor) WTF_REQUIRES_LOCK(g_vmListLock)
```
Nit: can be `const Invocable<IterationStatus(VM&)> auto& functor`

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -605,12 +605,8 @@ NO_RETURN_DUE_TO_CRASH NEVER_INLINE void MarkedBlock::dumpInfoAndCrashForInvalid
-            return IterationStatus::Done;
-        }
-        return IterationStatus::Continue;
+    isBlockVMValid = VMManager::findVM([&] (VM& vm) {
```
Maybe I should call this `findMatch` or `findMatchingVM` instead?

kmiller68 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -0,0 +1,91 @@
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+
+    template<typename Functor>
+    static VM* findVM(const Functor& test)
```
I would make this `const Invocable<TestCallBack> auto&` or inline the signature of `TestCallBack`.

kmiller68 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    template<typename Functor>
+    static void forEachVM(const Functor& functor)
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    template<typename Functor>
+    static Error forEachVMWithTimeout(Seconds timeout, const Functor& functor)
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/VMManager.h:
```diff
@@ -0,0 +1,91 @@
+    template<typename Functor>
+    static VM* findVM(const Functor& test)
+    {
+        return findVMImpl(scopedLambda<TestCallback, Functor>(test));
```
Do you need the `Functor` in the template list? Isn't it deduced by the first parameter? If you do need it you could change the above to `template<Invocable<TestCallBack> Functor>`

kmiller68 reviewed (approved):
r=me with some readability comments

hyjorc1 commented on Source/JavaScriptCore/runtime/VMManager.cpp:
```diff
@@ -0,0 +1,124 @@
+    VM* result = nullptr;
+    iterateVMs(scopedLambda<IteratorCallback>([&] (VM& vm) {
+        if (test(vm)) {
+            result = &vm;
```
Should we add this
```
s_recentVM = &vm;
```
here? Since then we can have
```
// Operation 1
VMManager::findVM([=](VM& vm) { return &vm == targetVM; });  // O(n) - cache miss

// Operation 2
VMManager::findVM([=](VM& vm) { return &vm == targetVM; });  // O(1) - cache hit!

// Operation 3
VMManager::findVM([=](VM& vm) { return &vm == targetVM; });  // O(1) - cache hit!
```

hyjorc1 reviewed (approved):
r=me if EWS is happy.

Commit:
Refactor VMManager out of VMInspector.
https://bugs.webkit.org/show_bug.cgi?id=297578
rdar://158659702

Reviewed by Keith Miller and Yijia Huang.

Previously, we only needed to enumerate VMs for debugging purposes.  As a result, we implemented
a container of VMs in VMInspector to support enumeration and inspection.  However, since then,
production code has come to rely on this container of VMs.  Additionally, we will soon need to
add more code in the area of VM management and coordination.  Hence, it makes sense to refactor
this container code out of VMInspector into a VMManager class of its own.

Changes include:

1. Changing VMManager::forEachVM() to take a ScopedLambda instead of a Function.  forEachVM() is
   an iteration mechanism, and the life cycle of the callback does not need to extend beyond the
   iteration operation.  Hence, then context of the iteration function can live on the stack in
   a ScopedLambda instead of needing to be malloc'ed in a Function.

2. Some clients were previously using VMInspector's VM iteration simply for the purpose of
   finding a VM matching some criteria.  Most of the time, the recent most added or searched
   for VM is the matching one.  Hence, we have a s_recentVM cache for that.  However, the
   iteration operation should still be holding the g_vmListLock to prevent the VM instance from
   being destructed before we call the client's callback with the VM in s_recentVM.

   Previously, this was done by requiring the client to manually acquire g_vmListLock before
   checking s_recentVM, and subsequently, using VMInspector::iterate() to iterate the VMs.
   This can be a source of bugs if the client does handle the lock correctly.  Additionally,
   it requires a lot more of VM container implementation to be exposed as API.

   In VMManager, we introduce a findVM() operation that will encapsulate this synchronization,
   checking of s_recentVM, and thereafter, iterating the container if needed.  This makes the
   interface more robust and simpler to use correctly.

3. Removed the dependency for VMInspector::isValidExecutableMemory() and
   VMInspector::codeBlockForMachinePC() on the VMInspector's lock.

   VMInspector::isValidExecutableMemory() never needed the VMInspector lock.  The real lock that
   actually needs to be acquired belongs to the ExecutableAllocator instead.
   VMInspector::isValidExecutableMemory() already acquires that.

   VMInspector::codeBlockForMachinePC() now uses VMManager::forEachVM() which handles its
   locking internally.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/disassembler/ARM64/A64DOpcode.cpp:
(JSC::ARM64Disassembler::A64DOpcodeMoveWide::handlePotentialDataPointer):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
* Source/JavaScriptCore/jsc.cpp:
(startTimeoutTimer):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/VMManager.cpp: Added.
(JSC::WTF_REQUIRES_LOCK):
(JSC::VMManager::add):
(JSC::VMManager::remove):
(JSC::VMManager::isValidVMSlow):
(JSC::VMManager::dumpVMs):
(JSC::VMManager::findVMImpl):
(JSC::VMManager::forEachVMImpl):
(JSC::VMManager::forEachVMWithTimeoutImpl):
* Source/JavaScriptCore/runtime/VMManager.h: Added.
(JSC::VMManager::isValidVM):
(JSC::VMManager::findVM):
(JSC::VMManager::forEachVM):
(JSC::VMManager::forEachVMWithTimeout):
* Source/JavaScriptCore/tools/HeapVerifier.cpp:
(JSC::HeapVerifier::checkIfRecorded):
* Source/JavaScriptCore/tools/Integrity.cpp:
(JSC::Integrity::Analyzer::analyzeVM):
* Source/JavaScriptCore/tools/IntegrityInlines.h:
(JSC::Integrity::doAudit):
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::vmForCallFrame):
(JSC::VMInspector::codeBlockForMachinePC):
(JSC::VMInspector::singleton): Deleted.
(JSC::VMInspector::add): Deleted.
(JSC::VMInspector::remove): Deleted.
(JSC::VMInspector::isValidVMSlow): Deleted.
(JSC::VMInspector::dumpVMs): Deleted.
(JSC::VMInspector::forEachVM): Deleted.
* Source/JavaScriptCore/tools/VMInspector.h:
(JSC::VMInspector::isValidVM): Deleted.
(JSC::VMInspector::WTF_RETURNS_LOCK): Deleted.
(JSC::VMInspector::WTF_REQUIRES_LOCK): Deleted.
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::reportToVMsForDestruction):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:

Canonical link: https://commits.webkit.org/298937@main
